pkg/cli/admin/prune/imageprune/prune.go

```diff
@@ -524,1083 +727,494 @@ func exceedsLimits(is *imagev1.ImageStream, image *imagev1.Image, limits map[str
-) (deletions []Deletion, failures []Failure) {
-	imgUpdateChan := p.imageWatcher.ResultChan()
-	isUpdateChan := p.imageStreamWatcher.ResultChan()
-	for {
-		// make workers busy
-		for len(p.processedImages) < p.numWorkers {
-			job, blocked := p.getNextJob()
-			if blocked {
-				break
-			}
-			if job == nil {
-				if len(p.processedImages) == 0 {
-					return
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for k := range workQueue {
+				mu.Lock()
+				stream := p.imageStreams[k]
+				mu.Unlock()
+
+				updatedStream, stats, errs := p.pruneImageStream(stream, streamPruner, layerLinkDeleter, manifestDeleter)
+
+				if updatedStream == nil {
+					mu.Lock()
+					delete(p.imageStreams, k)
+					mu.Unlock()
+				} else if updatedStream != stream {
+					mu.Lock()
+					p.imageStreams[k] = updatedStream
+					mu.Unlock()
 				}
-				break
-			}
-			jobChan <- job
-			p.processedImages[job.Image] = job
-		}
 
-		select {
-		case res := <-resultChan:
-			p.updateGraphWithResult(&res)
-			for _, deletion := range res.Deletions {
-				deletions = append(deletions, deletion)
-			}
-			for _, failure := range res.Failures {
-				failures = append(failures, failure)
-			}
-			delete(p.processedImages, res.Job.Image)
-		case <-isUpdateChan:
-			// TODO: fix gonum/graph to not reuse IDs of deleted nodes and reenable event handling
-			//p.handleImageStreamEvent(event)
-		case <-imgUpdateChan:
-			// TODO: fix gonum/graph to not reuse IDs of deleted nodes and reenable event handling
-			//p.handleImageEvent(event)
-		}
-	}
-}
+				pruneStats.Add(stats)
 
-// getNextJob removes a prunable image from the queue, makes a job out of it and returns it.
-// Image may be removed from the queue without being processed if it becomes not prunable (by being referred
-// by a new image stream). Image may also be skipped and processed later when it is currently blocked.
-//
-// Image is blocked when at least one of its components is currently being processed in a running job and
-// the component has either:
-//   - only one remaining strong reference from the blocked image (the other references are being currently
-//     removed)
-//   - only one remaining reference in an image stream, where the component is tagged (via image) (the other
-//     references are being currently removed)
-//
-// The concept of blocked images attempts to preserve image components until the very last image
-// referencing them is deleted. Otherwise an image previously considered as prunable becomes not prunable may
-// become not usable since its components have been removed already.
-func (p *pruner) getNextJob() (job *Job, blocked bool) {
-	if p.queue == nil {
-		return
-	}
-
-	pop := func(item *nodeItem) (*imagegraph.ImageNode, *nodeItem) {
-		node, next := item.pop()
-		if item == p.queue {
-			p.queue = next
-		}
-		return node, next
+				for _, err := range errs {
+					err = fmt.Errorf("imagestream %s/%s: %v", stream.Namespace, stream.Name, err)
+					klog.V(4).Info(err)
```

Comment:
it's nice to have them in debug logs, the caller will print the error only when the pruner finishes its job
in case of OOMKiller or panic this message might be crucial